identity/aziot-dps-client-async/src/lib.rs

```diff
@@ -13,323 +13,233 @@
-    let resource_uri = format!(
-        "/{}/registrations/{}/operations/{}?api-version=2018-11-01",
-        scope_id, registration_id, operation_id
-    );
-
-    let res: model::RegistrationOperationStatus = request::<(), _>(
-        uri,
-        scope_id,
-        registration_id,
-        http::Method::GET,
-        &resource_uri,
-        sas_key,
-        identity_cert,
-        identity_pk,
-        None,
-        key_client,
-        key_engine,
-        cert_client,
-    )
-    .await?;
-
-    Ok(res)
-}
+pub struct Client {
+    global_endpoint: String,
+    scope_id: String,
 
-#[derive(Debug, serde::Deserialize, serde::Serialize)]
-pub struct Error {
-    #[serde(alias = "Message")]
-    pub message: std::borrow::Cow<'static, str>,
+    key_client: Arc<aziot_key_client_async::Client>,
+    key_engine: Arc<futures_util::lock::Mutex<openssl2::FunctionalEngine>>,
+    cert_client: Arc<aziot_cert_client_async::Client>,
 }
 
-async fn request<TRequest, TResponse>(
-    global_endpoint: &str,
-    scope_id: &str,
-    registration_id: &str,
-    method: http::Method,
-    uri: &str,
-    sas_key: Option<String>,
-    identity_cert: Option<String>,
-    identity_pk: Option<String>,
-    body: Option<&TRequest>,
-    key_client: &aziot_key_client_async::Client,
-    key_engine: &mut openssl2::FunctionalEngineRef,
-    cert_client: &aziot_cert_client_async::Client,
-) -> std::io::Result<TResponse>
-where
-    TRequest: serde::Serialize,
-    TResponse: serde::de::DeserializeOwned,
-{
-    let uri = format!("{}{}", global_endpoint, uri);
-
-    let req = hyper::Request::builder().method(method).uri(uri);
-    // `req` is consumed by both branches, so this cannot be replaced with `Option::map_or_else`
-    //
-    // Ref: https://github.com/rust-lang/rust-clippy/issues/5822
-    #[allow(clippy::option_if_let_else)]
-    let req = if let Some(body) = body {
-        let body = serde_json::to_vec(body)
-            .expect("serializing request body to JSON cannot fail")
-            .into();
-        req.header(hyper::header::CONTENT_TYPE, "application/json")
-            .body(body)
-    } else {
-        req.body(hyper::Body::default())
-    };
-
-    let mut req = req.expect("cannot fail to create hyper request");
-
-    let connector = if let Some(key) = sas_key.clone() {
-        let (connector, token) =
-            get_sas_connector(scope_id.into(), registration_id.into(), key, key_client).await?;
-
-        let authorization_header_value = hyper::header::HeaderValue::from_str(&token)
-            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
-        req.headers_mut()
-            .append(hyper::header::AUTHORIZATION, authorization_header_value);
-        connector
-    } else {
-        get_x509_connector(
-            identity_cert.expect("device identity certificate not found"),
-            identity_pk.expect("device private key not found"),
+impl Client {
+    #[must_use]
```

Comment:
This is to satisfy `clippy::must_use_candidate` ? You can just `allow` it at the crate level, like `CONTRIBUTING.md` says.

Reply:
I won't lie, this `#[must_use]` is just copied/pasted from another Client implementation. It's nothing major, so I'll just go ahead and merge the PR.